index.d.ts

```diff
@@ -0,0 +1,49 @@
+/**
+ * html string helper function, flattens array values
+ */
+declare function html(
+  components: TemplateStringsArray,
+  ...values: (
+    | boolean
+    | void
+    | null
+    | string
+    | number
+    | (void | null | boolean | string | number)[])[]
+): string;
+
+/**
+ * create application store from reducer
+ */
+declare function createStore<State, Action>(
+  reducer: (state: State, action: Action) => State
+): {
+  /**
+   * dispatch an app action
+   */
+  dispatch(action: Action, ...args: any[]): void;
+
+  /**
+   * mount a component on an element
+   */
+  attach<Component extends (...args: any[]) => string>(
+    component: Component,
+    root: Element
+  ): void;
+
+  /**
+   * attach the app state to a component
+   */
+  connect(component: (state: State) => string): () => string;
+  connect<A>(component: (state: State, a: A) => string): (a: A) => string;
+  connect<A, B>(
+    component: (state: State, a: A, b: B) => string
+  ): (a: A, b: B) => string;
+  connect<A, B, C>(
+    component: (state: State, a: A, b: B, c: C) => string
+  ): (a: A, b: B, c: C) => string;
+  connect(component: (state: State, ...args: any[]) => string): (...args: any[]) => string;
```

Comment:
With this line here, are the other declarations using generics needed?

Reply:
This is sort of an unfortunate property of typescript lacking ["variadic kinds"](https://github.com/Microsoft/TypeScript/issues/5453), which is basically typed spread arguments.

If we only have the last signature, than any connected components will always have the function signature `(...args: any[]) => string` instead of specific argument types.

For example, say we have a component like the following

```typescript
const SayHi = (state: State, name: string) => html`
  <div>
    hey ${name}, check out this state prop ${ state.prop }
  </div>
`;
```

and then connect it

```typescript
const ConnectedSayHi = connect(SayHi);

// if we only have the last signature, this will not error,
// as the signature is (...args: any[]) => string
ConnectedSayHi(10);

// however, if we have the over loaded signature, 
// then connected components with 3 or fewer arguments 
// will have their arguments typechecked...
ConnectedSayHi(10) // type error, as the signature is (a: string) => string;
```

Reply:
the current annotation basically says... If the connected function has 3 or fewer arguments, create a new function with the same arguments (minus `state`), otherwise just type it as `(...args: any[]) => string`

Reply:
Thanks for the explanation. This makes sense to me now.